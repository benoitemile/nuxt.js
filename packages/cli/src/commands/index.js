export const start = () => import('./start')
export const dev = () => import('./dev')
export const build = () => import('./build')
export const generate = () => import('./generate')
export const help = () => import('./help')
