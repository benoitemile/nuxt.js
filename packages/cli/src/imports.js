import path from 'path'

const localNodeModules = path.resolve(process.cwd(), 'node_modules')

// Prefer importing modules from local node_modules (for NPX and global bin)
async function _import(modulePath) {
  let m
  try {
    m = await import(path.resolve(localNodeModules, modulePath))
  } catch (e) {
    m = await import(modulePath)
  }
  return m
}

export const builder = () => _import('@nuxt/builder')
export const webpack = () => _import('@nuxt/webpack')
export const generator = () => _import('@nuxt/generator')
export const core = () => _import('@nuxt/core')
