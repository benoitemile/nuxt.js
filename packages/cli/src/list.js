import chalk from 'chalk'
import NuxtCommand from './command'
import { indent, foldLines, startSpaces, optionSpaces, colorize } from './utils/formatting'

export default async function listCommands() {
  const commandsOrder = ['dev', 'build', 'generate', 'start', 'help']

  // Load all commands
  const _commands = await Promise.all(
    commandsOrder.map(cmd => NuxtCommand.load(cmd))
  )

  let maxLength = 0
  const commandsHelp = []

  for (const name in _commands) {
    commandsHelp.push([_commands[name].usage, _commands[name].description])
    maxLength = Math.max(maxLength, _commands[name].usage.length)
  }

  const _cmmds = commandsHelp.map(([cmd, description]) => {
    const i = indent(maxLength + optionSpaces - cmd.length)
    return foldLines(
      chalk.green(cmd) + i + description,
      startSpaces + maxLength + optionSpaces * 2,
      startSpaces + optionSpaces
    )
  }).join('\n')

  const usage = foldLines(`Usage: nuxt <command> [--help|-h]`, startSpaces)
  const cmmds = foldLines(`Commands:`, startSpaces) + '\n\n' + _cmmds

  process.stderr.write(colorize(`${usage}\n\n${cmmds}\n\n`))
}
