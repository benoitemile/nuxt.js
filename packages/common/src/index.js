export { default as Hookable } from './hookable'
export { default as ModernBrowsers } from '../data/modern-browsers.json'
export * from './utils'
