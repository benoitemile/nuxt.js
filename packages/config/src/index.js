export { getDefaultNuxtConfig } from './config'
export { getNuxtConfig } from './options'
