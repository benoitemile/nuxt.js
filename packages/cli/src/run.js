import consola from 'consola'
import NuxtCommand from './command'
import * as commands from './commands'
import setup from './setup'
import listCommands from './list'

export default function run() {
  const defaultCommand = 'dev'
  let cmd = process.argv[2]

  if (commands[cmd]) { // eslint-disable-line import/namespace
    process.argv.splice(2, 1)
  } else {
    if (process.argv.includes('--help') || process.argv.includes('-h')) {
      listCommands().then(() => process.exit(0))
      return
    }
    cmd = defaultCommand
  }

  // Setup runtime
  setup({
    dev: cmd === 'dev'
  })

  return NuxtCommand.load(cmd)
    .then(command => command.run())
    .catch((error) => {
      consola.fatal(error)
    })
}
