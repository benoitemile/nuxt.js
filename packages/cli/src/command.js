import parseArgs from 'minimist'
import { name, version } from '../package.json'
import { loadNuxtConfig } from './utils'
import { indent, foldLines, startSpaces, optionSpaces, colorize } from './utils/formatting'
import * as commands from './commands'
import * as imports from './imports'

export default class NuxtCommand {
  constructor({ name, description, usage, options, run } = {}) {
    this.name = name || ''
    this.description = description || ''
    this.usage = usage || ''
    this.options = Object.assign({}, options)
    this._run = run
  }

  static async load(name) {
    if (name in commands) {
      const cmd = await commands[name]() // eslint-disable-line import/namespace
        .then(m => m.default)
      return NuxtCommand.from(cmd)
    } else {
      // TODO dynamic module loading
      throw new Error('Command ' + name + ' could not be loaded!')
    }
  }

  static from(options) {
    if (options instanceof NuxtCommand) {
      return options
    }
    return new NuxtCommand(options)
  }

  run() {
    return this._run(this)
  }

  showVersion() {
    process.stdout.write(`${name} v${version}\n`)
    process.exit(0)
  }

  showHelp() {
    process.stdout.write(this._getHelp())
    process.exit(0)
  }

  getArgv(args) {
    const minimistOptions = this._getMinimistOptions()
    const argv = parseArgs(args || process.argv.slice(2), minimistOptions)

    if (argv.version) {
      this.showVersion()
    } else if (argv.help) {
      this.showHelp()
    }

    return argv
  }

  async getNuxtConfig(argv, extraOptions) {
    const config = await loadNuxtConfig(argv)
    const options = Object.assign(config, extraOptions || {})

    for (const name of Object.keys(this.options)) {
      this.options[name].prepare && this.options[name].prepare(this, options, argv)
    }

    return options
  }

  async getNuxt(options) {
    const { Nuxt } = await imports.core()
    return new Nuxt(options)
  }

  async getBuilder(nuxt) {
    const { Builder } = await imports.builder()
    const { BundleBuilder } = await imports.webpack()
    return new Builder(nuxt, BundleBuilder)
  }

  async getGenerator(nuxt) {
    const { Generator } = await imports.generator()
    const builder = await this.getBuilder(nuxt)
    return new Generator(nuxt, builder)
  }

  _getMinimistOptions() {
    const minimistOptions = {
      alias: {},
      boolean: [],
      string: [],
      default: {}
    }

    for (const name of Object.keys(this.options)) {
      const option = this.options[name]

      if (option.alias) {
        minimistOptions.alias[option.alias] = name
      }
      if (option.type) {
        minimistOptions[option.type].push(option.alias || name)
      }
      if (option.default) {
        minimistOptions.default[option.alias || name] = option.default
      }
    }

    return minimistOptions
  }

  _getHelp() {
    const options = []
    let maxOptionLength = 0

    for (const name in this.options) {
      const option = this.options[name]

      let optionHelp = '--'
      optionHelp += option.type === 'boolean' && option.default ? 'no-' : ''
      optionHelp += name
      if (option.alias) {
        optionHelp += `, -${option.alias}`
      }

      maxOptionLength = Math.max(maxOptionLength, optionHelp.length)
      options.push([ optionHelp, option.description ])
    }

    const _opts = options.map(([option, description]) => {
      const i = indent(maxOptionLength + optionSpaces - option.length)
      return foldLines(
        option + i + description,
        startSpaces + maxOptionLength + optionSpaces * 2,
        startSpaces + optionSpaces
      )
    }).join('\n')

    const usage = foldLines(`Usage: nuxt ${this.usage} [options]`, startSpaces)
    const description = foldLines(this.description, startSpaces)
    const opts = foldLines(`Options:`, startSpaces) + '\n\n' + _opts

    let helpText = colorize(`${usage}\n\n`)
    if (this.description) helpText += colorize(`${description}\n\n`)
    if (options.length) helpText += colorize(`${opts}\n\n`)

    return helpText
  }
}
