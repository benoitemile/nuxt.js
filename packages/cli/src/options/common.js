import { defaultNuxtConfigFile } from '@nuxt/config'
import { normalizeArg } from '../utils'

export default {
  spa: {
    alias: 's',
    type: 'boolean',
    description: 'Launch in SPA mode'
  },
  universal: {
    alias: 'u',
    type: 'boolean',
    description: 'Launch in Universal mode (default)'
  },
  'config-file': {
    alias: 'c',
    type: 'string',
    default: defaultNuxtConfigFile,
    description: `Path to Nuxt.js config file (default: \`${defaultNuxtConfigFile}\`)`
  },
  modern: {
    alias: 'm',
    type: 'string',
    description: 'Build/Start app for modern browsers, e.g. server, client and false',
    prepare(cmd, options, argv) {
      if (argv.modern !== undefined) {
        options.modern = normalizeArg(argv.modern)
      }
    }
  },
  // TODO: Change this to default: true in Nuxt 3
  'force-exit': {
    type: 'boolean',
    default: false,
    description: 'Force Nuxt.js to exit after the command has finished (this option has no effect on commands which start a server)'
  },
  version: {
    alias: 'v',
    type: 'boolean',
    description: 'Display the Nuxt version'
  },
  help: {
    alias: 'h',
    type: 'boolean',
    description: 'Display this message'
  }
}
