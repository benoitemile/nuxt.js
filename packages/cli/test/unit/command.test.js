import Command from '../../src/command'
import { common, server } from '../../src/options'
import { consola } from '../utils'

jest.mock('@nuxt/core')
jest.mock('@nuxt/builder')
jest.mock('@nuxt/generator')

const allOptions = {
  ...common,
  ...server
}

describe('cli/command', () => {
  beforeEach(() => jest.restoreAllMocks())

  test('builds minimist options', () => {
    const cmd = new Command({ options: allOptions })
    const minimistOptions = cmd._getMinimistOptions()

    expect(minimistOptions.string.length).toBe(5)
    expect(minimistOptions.boolean.length).toBe(4)
    expect(minimistOptions.alias.c).toBe('config-file')
    expect(minimistOptions.default.c).toBe(common['config-file'].default)
  })

  test('parses args', () => {
    const cmd = new Command({ options: { ...common, ...server } })

    let args = ['-c', 'test-file', '-s', '-p', '3001']
    let argv = cmd.getArgv(args)

    expect(argv['config-file']).toBe(args[1])
    expect(argv.spa).toBe(true)
    expect(argv.universal).toBe(false)
    expect(argv.port).toBe('3001')

    args = ['--no-build']
    argv = cmd.getArgv(args)

    expect(argv.build).toBe(false)
  })

  test('prints version automatically', () => {
    const cmd = new Command()
    cmd.showVersion = jest.fn()

    const args = ['--version']
    cmd.getArgv(args)

    expect(cmd.showVersion).toHaveBeenCalledTimes(1)
  })

  test('prints help automatically', () => {
    const cmd = new Command({ options: allOptions })
    cmd.showHelp = jest.fn()

    const args = ['-h']
    cmd.getArgv(args)

    expect(cmd.showHelp).toHaveBeenCalledTimes(1)
  })

  test('returns nuxt config', async () => {
    const cmd = new Command({ options: allOptions })

    const args = ['-c', 'test-file', '-a', '-p', '3001', '-q', '-H']
    const argv = cmd.getArgv(args)
    argv._ = ['.']

    const options = await cmd.getNuxtConfig(argv, { testOption: true })

    expect(options.testOption).toBe(true)
    expect(options.server.port).toBe(3001)
    expect(consola.fatal).toHaveBeenCalledWith('Provided hostname argument has no value') // hostname check
  })

  test('returns Nuxt instance', async () => {
    const cmd = new Command()
    const nuxt = await cmd.getNuxt()

    expect(nuxt.constructor.name).toBe('Nuxt')
    expect(typeof nuxt.ready).toBe('function')
  })

  test('returns Builder instance', async () => {
    const cmd = new Command()
    const builder = await cmd.getBuilder()

    expect(builder.constructor.name).toBe('Builder')
    expect(typeof builder.build).toBe('function')
  })

  test('returns Generator instance', async () => {
    const cmd = new Command()
    const generator = await cmd.getGenerator()

    expect(generator.constructor.name).toBe('Generator')
    expect(typeof generator.generate).toBe('function')
  })

  test('builds help text', () => {
    const cmd = new Command({
      description: 'a very long description that should not wrap to the next line because is not longer ' +
        'than the terminal width',
      usage: 'this is how you do it',
      options: {
        ...allOptions,
        foo: {
          type: 'boolean',
          description: 'very long option that is not longer than the terminal width and ' +
        'should not wrap to the next line'
        }
      }
    })

    expect(cmd._getHelp()).toMatchSnapshot()
  })

  test('loads command from name', async () => {
    const cmd = await Command.load('dev')
    expect(cmd._getHelp()).toMatchSnapshot()
  })

  test('show version prints to stdout and exits', () => {
    jest.spyOn(process.stdout, 'write').mockImplementation(() => {})
    jest.spyOn(process, 'exit').mockImplementationOnce(code => code)

    const cmd = new Command()
    cmd.showVersion()

    expect(process.stdout.write).toHaveBeenCalled()
    expect(process.exit).toHaveBeenCalled()

    process.stdout.write.mockRestore()
    process.exit.mockRestore()
  })

  test('show help prints to stdout and exits', () => {
    jest.spyOn(process.stdout, 'write').mockImplementation(() => {})
    jest.spyOn(process, 'exit').mockImplementationOnce(code => code)

    const cmd = new Command()
    cmd.showHelp()

    expect(process.stdout.write).toHaveBeenCalled()
    expect(process.exit).toHaveBeenCalled()

    process.stdout.write.mockRestore()
    process.exit.mockRestore()
  })
})
