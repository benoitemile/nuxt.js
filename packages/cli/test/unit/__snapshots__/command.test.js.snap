// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`cli/command builds help text 1`] = `
"  Usage: nuxt this is how you do it [options]

  a very long description that should not wrap to the next line because is not longer than the terminal width

  Options:

    --spa, -s          Launch in SPA mode
    --universal, -u    Launch in Universal mode (default)
    --config-file, -c  Path to Nuxt.js config file (default: nuxt.config.js)
    --modern, -m       Build/Start app for modern browsers, e.g. server, client and false
    --version          Display the Nuxt version
    --help, -h         Display this message
    --port, -p         Port number on which to start the application
    --hostname, -H     Hostname on which to start the application
    --unix-socket, -n  Path to a UNIX socket
    --foo              very long option that is not longer than the terminal width and should not wrap to the next line

"
`;

exports[`cli/command loads command from name 1`] = `
"  Usage: nuxt dev <dir> [options]

  Start the application in development mode (e.g. hot-code reloading, error reporting)

  Options:

    --spa, -s          Launch in SPA mode
    --universal, -u    Launch in Universal mode (default)
    --config-file, -c  Path to Nuxt.js config file (default: nuxt.config.js)
    --modern, -m       Build/Start app for modern browsers, e.g. server, client and false
    --version          Display the Nuxt version
    --help, -h         Display this message
    --port, -p         Port number on which to start the application
    --hostname, -H     Hostname on which to start the application
    --unix-socket, -n  Path to a UNIX socket

"
`;
